src/messages/textDocument_document.cc

```diff
@@ -165,88 +152,70 @@ void MessageHandler::textDocument_documentSymbol(JsonReader &reader,
+
+    /**
+     * with 2 ranges that start at the same Position, we want the wider one
+     * first (swap lhs/rhs)
+     *
+     * we use range for start to start at symbol name
+     * issue with: int i, j, k; otherwise
+     */
+    auto sym_cmp = [](ExtentRef const &lhs, ExtentRef const &rhs) {
+      return lhs.range.start < rhs.range.start ||
+             (lhs.range.start == rhs.range.start &&
+              rhs.extent.end < lhs.extent.end);
+    };
+
     for (auto [sym, refcnt] : file->symbol2refcnt) {
       if (refcnt <= 0 || !sym.extent.valid())
         continue;
-      auto r = sym2ds.try_emplace(SymbolIdx{sym.usr, sym.kind});
-      auto &ds = r.first->second;
-      if (!ds || sym.role & Role::Definition) {
-        if (!ds)
-          ds = std::make_unique<DocumentSymbol>();
-        if (auto range = getLsRange(wf, sym.range)) {
-          ds->selectionRange = *range;
-          ds->range = ds->selectionRange;
-          // For a macro expansion, M(name), we may use `M` for extent and
-          // `name` for spell, do the check as selectionRange must be a subrange
-          // of range.
-          if (sym.extent.valid())
-            if (auto range1 = getLsRange(wf, sym.extent);
-                range1 && range1->includes(*range))
-              ds->range = *range1;
-        }
+      syms.push_back(sym);
+    }
+    std::sort(std::begin(syms), std::end(syms), sym_cmp);
+
+    std::vector<DocumentSymbol> result;
+    std::stack<DocumentSymbol *> indent;
+    for (auto sym : syms) {
+      DocumentSymbol ds;
+      if (auto range = getLsRange(wf, sym.range)) {
+        ds.selectionRange = *range;
+        ds.range = ds.selectionRange;
+        // For a macro expansion, M(name), we may use `M` for extent and
+        // `name` for spell, do the check as selectionRange must be a subrange
+        // of range.
+        if (sym.extent.valid())
+          if (auto range1 = getLsRange(wf, sym.extent);
+              range1 && range1->includes(*range))
+            ds.range = *range1;
       }
-      if (!r.second)
-        continue;
-      std::vector<const void *> def_ptrs;
-      SymbolKind kind = SymbolKind::Unknown;
       withEntity(db, sym, [&](const auto &entity) {
-        auto *def = entity.anyDef();
+        auto const *def = entity.anyDef();
         if (!def)
           return;
-        ds->name = def->name(false);
-        ds->detail = def->detailed_name;
-        for (auto &def : entity.def)
-          if (def.file_id == file_id && !ignore(&def)) {
-            kind = ds->kind = def.kind;
-            def_ptrs.push_back(&def);
+        ds.name = def->name(false);
+        ds.detail = def->detailed_name;
+        ds.kind = def->kind;
+
+        if (!ignore(def) && (ds.kind == SymbolKind::Namespace || allows(sym))) {
+          // drop symbols that are behind the current one
+          while (!indent.empty() &&
+                 // use selectionRange for start to start at symbol name
```

Comment:
Elaborate this comment a bit? How does `int i, j, k;` cause an issue?

Reply:
When we define multiple variable on "the same line" eg: int i, j, k;
k is child of j and j is child of i (if we use `range`)

This is because `range` starts at the type so at the same position for i, j and k, but with `selectionRange` it starts at the name of the variable thus we can separate the 3 and not make j a child of i.

The outline before:
k
 j
  i
after:
i
j
k

Reply:
I don't understand. With the current `while (!scopes.empty() && scopes.back()->range.end <= ds->selectionRange.start)`, i is the parent of j and j is the parent of k. Only by switching to `ds->range.start` we can make them flattened.

Reply:
`ds->range` is `sym.extent` if valid (it's a little bit confusing but that's how it was) and sym.extent begins at the type so here: `int i, j, k;`
with `ds->selectionRange` (which is `sym.range`) it begins at i or j or k

I don't fully understand where we disagree